src/ansi-c/c_typecheck_type.cpp

```diff
@@ -1160,10 +1160,20 @@ void c_typecheck_baset::typecheck_c_enum_type(typet &type)
     throw 0;
   }
 
-  if(as_expr.find(ID_enum_underlying_type).is_not_nil())
+  const bool have_underlying_type =
+    type.find_type(ID_enum_underlying_type).is_not_nil();
+
+  if(have_underlying_type)
   {
-    warning().source_location = source_location;
-    warning() << "ignoring specification of underlying type for enum" << eom;
+    typecheck_type(type.add_type(ID_enum_underlying_type));
+
+    if(!is_signed_or_unsigned_bitvector(
+         static_cast<const typet &>(type.find(ID_enum_underlying_type))))
+    {
+      error().source_location = source_location;
+      error() << "underlying type for enum must be an integral type" << eom;
+      throw 0;
```

Comment:
Do we not have a structured exception to throw here? (and in the other cases below)

Reply:
Yes this should an `invalid_source_file_exceptiont`

Reply:
Done